### ISSUE TITLE
load_model issue

### ISSUE THREAD

**Issue by chuong98** [TextpertAi/alt-model-checkpoint#1]
Hi, thanks for the package. 
I just have the problem when I resume the training, using load_model. Apparently, I must recompile the model again, a.k.a:

```
with tf.device('/cpu:0'):
        based_model = load_model(filepath=model_path, custom_objects={'AnchorBoxes': AnchorBoxes,
                                                       'L2Normalization': L2Normalization,
                                                       'compute_loss': ssd_loss.compute_loss})
based_model.summary()
model = multi_gpu_model(based_model, gpus=2)   
 # model.compile(optimizer=opt, loss=ssd_loss.compute_loss)
```
where based_model is the model saved by alt-model-checkpoint. 
Without recompiling, it throws error.
If I recompile, it can retrieve the model config and the weight, but I will lose the optimizer stage. Do you known how to resume the training with a full model (model_config + weights + optimizer ). Thank you.

**Comment by evictor** [comment]
Is the plan to resume training with the multi-gpu model? I think in this case you might want to _also_ use the standard `ModelCheckpoint(...)` with the multi-gpu model, then load that saved model for resumed training.

**Comment by nicolefinnie** [comment]
@evictor  I hit the same problem, the only difference is I did not have any custom objects as the original post did. I output the optimizer and the optimizer weights and updates are saved but I think they're not saved properly, because after I resumed the training, the accuracy kept going down and the loss kept going up, it's a classic symptom of the saved optimizer not learned from the previous model. 

What do you mean by "using the standard `ModelCheckPoint()`? Once the checkpoint is saved through your code, the checkpoint is saved as a normal `h5` model.  So you have to load the checkpoint as usual, pass it to `multiple_gpu_model` and compile this multiple_gpu_model with the saved `optimizer, loss, metrics, loss_weights, weighted_metrics`, you name it.  However, even if I resumed training with just one GPU using this saved model, the results are the same.

Something else I also noticed is the loss saved in the model is way higher when the model gets loaded again.  I wonder if you have encountered this problem? This code might not work for stateful optimizers.

**Comment by evictor** [comment]
Thanks for your feedback. The motivation behind AltModelCheckpoint is to save a model that is more robust for later inference or ad hoc use and not tied to the original multi-gpu architecture on which its weights were trained. For this use case optimizer state is not important as the model is not trained further in those scenarios.

When I say "use the standard ModelCheckpoint", I mean you could feasibly use _both_ AltModelCheckpoint _and_ ModelCheckpoint to accomplish both goals of 1) saving a "basic" model for inference (AltModelCheckpoint) and 2) saving your multi-gpu model exactly as is for training continuation later, e.g.:

```python
base_model = Model(...)
gpu_model = multi_gpu_model(base_model)
gpu_model.compile(...)

gpu_model.fit(..., callbacks=[
    AltModelCheckpoint('save/path/for/base_model.hdf5', base_model),
    ModelCheckpoint('save/path/for/multigpu_model.hdf5'),
])
```

In this case you should be able to resume training from `save/path/for/multigpu_model.hdf5` without losing optimizer state, but also use the trained `save/path/for/base_model.hdf5` for inference on different systems (CPU only, for example).

**Comment by nicolefinnie** [comment]
@evictor  Thanks for your response. However, this is the whole point to use a solution like yours, because you cannot save the model check point when you `fit` a `multiple_gpu_model`. There are some half-solutions such as [saving checkpoint's weights only](https://github.com/keras-team/keras/issues/2436#issuecomment-354882296).  However, this is an unsolved issue in Keras.

**Comment by evictor** [comment]
> you cannot save the model check point when you `fit` a `multiple_gpu_model`

Why not? That is what `ModelCheckpoint` will do. Internally it calls `model.save` which by default saves model architecture, weights, and optimizer state.

If you want to resume training on a _different_ system architecture entirely but inherit the previous optimizer state I understand that is a limitation with existing solutions, but it shouldn't be too difficult to build that. Is that what you're looking for?

**Comment by nicolefinnie** [comment]
@evictor, That's the point, you will hit problems when a `multi_gpu_model` instance calls `model.save`. It's been a pain in the ass for the Keras team for a long time. See [issue 8764](https://github.com/keras-team/keras/issues/8764) or [issue 8123](https://github.com/keras-team/keras/issues/8123). In general, Francois Chollet and the Keras team recommend people saving the template model (non-multiple-gpu-model).  The only thing that would work is to save the weights only via `ModelCheckPoint` in the `multi_gpu_model` case, where you lose optimizer states.

You can even add this [patch](https://stackoverflow.com/questions/47210811/can-not-save-model-using-model-save-following-multi-gpu-model-in-keras/48066771#48066771) to work around the bug with `multi-gpu-model`. 

And to your question, no, I'd like to resume training on the same system and in the same Keras version. However, I've written a workaround, thanks for your response anyway.

**Comment by evictor** [comment]
Ah, I see, thanks for explaining + refs. I'm surprised nobody has pushed that fix to Keras core.

**Comment by divyanipatil** [comment]
I tried compiling base_model alongwith the multi_gpu_model and used AltModelCheckpoint to save the base_model. So, later when I load the model, it doesn't throw any errors. But, I'm not sure if the optimizer state of the base_model that's saved matches with the gpu_model.
Is the optimizer state shared between the models as are the weights?

**Comment by evictor** [comment]
If you are using it with `save_weights_only=False` (the default) then optimizer state _will_ be saved for the model being saved (i.e. `alternate_model`).

I can't say for certain whether the alt model optimizer will match the multi-gpu model—that question would be answered in `compile()` logic in Keras, i.e. whether the optimizer on the multi-gpu model is referenced within submodels. I would doubt that it is as the submodels aren't optimizing.

A stopgap solution would be to _also_ use `ModelCheckpoint` in the multi-gpu model callbacks; this will ensure the multi-gpu model optimizer gets saved. You will have to do surgery to stitch together the optimizer from multi-gpu model into your saved base model though.

These are unusual use cases IMO. The only time you'd want to do this is if you wanted to resume training a model on a completely different system configuration. Usually you train on one configuration, then predict on another configuration. The optimizer state doesn't matter at inference time.

Give me a bit and I will find a way to weave in saving of the optimizer for the multi-gpu model in this helper class.

**Comment by evictor** [comment]
I have just pushed a new version that will inherit optimizer state from the base model when saving the alternate model. I believe this will solve everyone's problems in this thread. Please LMK but closing this now.

The new version is a major version bump to 2.0.0 because behavior is now changed by default to inherit the optimizer state. 👍

**Comment by divyanipatil** [comment]
@evictor, just what was needed. Thank you so much for this update :)
Though I had one question: Why do we need 'inherit_optimizer' as an extra param? Can't we simply use the value of 'save_weights_only' param from ModelCheckpoint instead??
As of now, we need to keep track of two params - resuming training is possible only when inherit_optimizer is True and save_weights_only is False. In the case where both are False --> resume is not possible as optmizer state will be saved but it won't match that of the trained model. 
And in rest of cases where save_weights_only is True, value of inherit_optimizer doesn't matter as only weights are saved when super().on_epoch_end() is called.
Any specific reason as to why inherit_optmizer was added?

**Comment by evictor** [comment]
> Can't we simply use the value of 'save_weights_only' param from ModelCheckpoint instead?

Keras interface here is a bit confusing. Short answer is no; `save_weights_only` actually omits the optimizer and other aspects of the model (architectural metadata, etc.). It is only for layer weights.

As far as keeping track of params, you should be able to use 2.0.0 of this lib with default param settings (the defaults are `inherit_optimizer=True, save_weights_only=False`) and achieve what you want. The resulting model file will contain 1) architecture of the "alternate model" (the template model, the pre-multi-GPU model, however you want to call it), 2) trained layer weights, _and_ 3) optimizer and other training state. *You shouldn't need to tweak other params.*

> In the case where both are False --> resume is not possible as optmizer state will be saved but it won't match that of the trained model.

That's true but I don't know why you would set both those to False. :) The point of inheriting optimizer prior to saving is to be able to resume training against the saved underlying model.

> And in rest of cases where save_weights_only is True, value of inherit_optimizer doesn't matter as only weights are saved when super().on_epoch_end() is called.

Again this is true, but I don't know why you would do this.

> Any specific reason as to why inherit_optmizer was added?

Just to explicitize behavior that might not otherwise be obvious with the ability to switch it off if needed for some reason. The defaults will do what this lib was designed for, but maybe there is some weird edge case where someone wants to save optimizer state from the underlying models.

**Comment by divyanipatil** [comment]
Thanks for the explanation!
